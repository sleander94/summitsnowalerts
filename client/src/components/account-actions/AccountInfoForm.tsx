import { useEffect, useState, FormEvent, KeyboardEvent } from 'react';
import { Navigate } from 'react-router-dom';
import { AuthProps } from '../../types.d';

const AccountInfoForm = ({ user }: AuthProps) => {
  const getAccountInfo = async () => {
    try {
      const response = await fetch('/users', {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      });
      const data = await response.json();
      setName(data.name);
      setEmail(data.email);
      setNewEmail(data.email);
      setPhone(data.phone);
      setEmailAlert(data.emailAlert);
      setTextAlert(data.textAlert);
      setMountains(data.mountains);
    } catch (err) {
      console.error(err);
    }
  };

  useEffect(() => {
    getAccountInfo();
  }, [user]);

  const [email, setEmail] = useState<string>();
  const [password, setPassword] = useState<string>('');
  const [mountains, setMountains] = useState<Object>();

  const [name, setName] = useState<string>();
  const [newEmail, setNewEmail] = useState<string>();
  const [phone, setPhone] = useState<string>();
  const [newPass, setNewPass] = useState<string>('');
  const [confirmPass, setConfirmPass] = useState<string>('');
  const [emailAlert, setEmailAlert] = useState<boolean>(false);
  const [textAlert, setTextAlert] = useState<boolean>(false);

  const [editName, setEditName] = useState<boolean>(false);
  const [editEmail, setEditEmail] = useState<boolean>(false);
  const [editPhone, setEditPhone] = useState<boolean>(false);
  const [editPass, setEditPass] = useState<boolean>(false);

  const [nameErr, setNameErr] = useState<string>();
  const [emailErr, setEmailErr] = useState<string>();
  const [phoneErr, setPhoneErr] = useState<string>();
  const [passwordErr, setPasswordErr] = useState<string>();
  const [confirmPassErr, setConfirmPassErr] = useState<string>();

  // Set hooks to watch state of each form value & throw appropriate errors
  useEffect(() => {
    name === '' ? setNameErr('Enter a name.') : setNameErr(undefined);
  }, [name]);

  useEffect(() => {
    !newEmail?.match(/([a-zA-Z0-9+._-]+@[a-zA-Z0-9._-]+\.[a-zA-Z0-9_-]+)/gi)
      ? setEmailErr('Enter a valid email.')
      : setEmailErr(undefined);
  }, [newEmail]);

  useEffect(() => {
    phone?.match(/[a-zA-Z]/gi)
      ? setPhoneErr('Phone number cannot contain letters.')
      : setPhoneErr(undefined);
  }, [phone]);

  useEffect(() => {
    if (newPass || confirmPass) {
      !newPass?.match(/^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{6,}$/gi)
        ? setPasswordErr('6+ characters with 1 letter and 1 number.')
        : setPasswordErr(undefined);
      newPass !== confirmPass
        ? setConfirmPassErr("Password don't match.")
        : setConfirmPassErr(undefined);
    }
  }, [newPass, confirmPass]);

  // Set general error value to true if any form value throws an error
  const [error, setError] = useState<boolean>(false);

  useEffect(() => {
    if (nameErr || emailErr || phoneErr || passwordErr || confirmPassErr)
      setError(true);
    else setError(false);
  }, [nameErr, emailErr, phoneErr, passwordErr, confirmPassErr]);

  const [posting, setPosting] = useState<boolean>(false);
  const [posted, setPosted] = useState<boolean>(false);
  const [postError, setPostError] = useState<string>();

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setPosted(false);
    if (!error) {
      try {
        console.log('posting');
        setPosting(true);
        const response = await fetch('/users', {
          method: 'PUT',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            name,
            email,
            newEmail,
            phone,
            newPass,
            emailAlert,
            textAlert,
            password,
            mountains,
          }),
        });
        console.log(response);
        if (response.ok) {
          console.log('updated');
          setPostError(undefined);
          setPosted(true);
          setPassword('');
        } else if (response.status === 401) {
          setPostError('Incorrect password.');
        } else {
          setPostError('Error updating account. Please try again.');
        }
        setPosting(false);
      } catch (err) {
        console.error(err);
      }
    } else {
      setPostError("Can't submit form with errors. ");
    }
  };

  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key == 'Enter') {
      const target = e.target as HTMLInputElement;
      target.blur();
    }
  };

  return (
    <section id="account-info-form">
      {!user && <Navigate to="/" />}
      <h1>Account Info</h1>
      <form action="" onSubmit={(e) => handleSubmit(e)}>
        <div className="form-field">
          <label htmlFor="name">Name: </label>
          {!editName && (
            <>
              <p>{name}</p>
              <button type="button" onClick={() => setEditName(true)}>
                <img
                  src={require('../../assets/icons/icons8-edit.svg').default}
                ></img>
              </button>
            </>
          )}
          {editName && (
            <>
              <input
                type="text"
                name="name"
                autoFocus
                value={name}
                onChange={(e) => {
                  e.target.className = 'touched';
                  setName(e.target.value);
                }}
                onBlur={() => (!nameErr ? setEditName(false) : null)}
                onKeyDown={(e) => handleKeyDown(e)}
              />
              {nameErr && <p className="error">{nameErr}</p>}
              <button
                type="button"
                onClick={() => (!nameErr ? setEditName(false) : null)}
              >
                <img
                  src={require('../../assets/icons/icons8-done.svg').default}
                ></img>
              </button>
            </>
          )}
        </div>
        <div className="form-field">
          <label htmlFor="email">Email: </label>
          {!editEmail && (
            <>
              <p>{newEmail}</p>
              <button type="button" onClick={() => setEditEmail(true)}>
                <img
                  src={require('../../assets/icons/icons8-edit.svg').default}
                ></img>
              </button>
            </>
          )}
          {editEmail && (
            <>
              <input
                type="email"
                name="email"
                autoFocus
                value={newEmail}
                onChange={(e) => {
                  e.target.className = 'touched';
                  setNewEmail(e.target.value);
                }}
                onBlur={() => (!emailErr ? setEditEmail(false) : null)}
                onKeyDown={(e) => handleKeyDown(e)}
              />
              {emailErr && <p className="error">{emailErr}</p>}
              <button
                type="button"
                onClick={() => (!emailErr ? setEditEmail(false) : null)}
              >
                <img
                  src={require('../../assets/icons/icons8-done.svg').default}
                ></img>
              </button>
            </>
          )}
        </div>
        <div className="form-field">
          <label htmlFor="phone">Phone: </label>
          {!editPhone && (
            <>
              <p>{phone}</p>
              <button type="button" onClick={() => setEditPhone(true)}>
                <img
                  src={require('../../assets/icons/icons8-edit.svg').default}
                ></img>
              </button>
            </>
          )}
          {editPhone && (
            <>
              <input
                type="tel"
                name="phone"
                autoFocus
                value={phone}
                onChange={(e) => {
                  e.target.className = 'touched';
                  setPhone(e.target.value);
                }}
                onBlur={() => (!phoneErr ? setEditPhone(false) : null)}
                onKeyDown={(e) => handleKeyDown(e)}
              />
              {phoneErr && <p className="error">{phoneErr}</p>}
              <button
                type="button"
                onClick={() => (!phoneErr ? setEditPhone(false) : null)}
              >
                <img
                  src={require('../../assets/icons/icons8-done.svg').default}
                ></img>
              </button>
            </>
          )}
        </div>
        <div className="form-field">
          <label htmlFor="newPass">Password: </label>
          {!editPass && (
            <>
              <p>{newPass ? '*'.repeat(newPass.length) : '******'}</p>
              <button type="button" onClick={() => setEditPass(true)}>
                <img
                  src={require('../../assets/icons/icons8-edit.svg').default}
                ></img>
              </button>
            </>
          )}
          {editPass && (
            <>
              <input
                type="password"
                name="newPass"
                autoFocus
                value={newPass}
                onChange={(e) => {
                  e.target.className = 'touched';
                  setNewPass(e.target.value);
                }}
                onKeyDown={(e) => handleKeyDown(e)}
              />
              {passwordErr && <p className="error">{passwordErr}</p>}
              <button
                type="button"
                onClick={() =>
                  !passwordErr && !confirmPassErr ? setEditPass(false) : null
                }
              >
                <img
                  src={require('../../assets/icons/icons8-done.svg').default}
                ></img>
              </button>
            </>
          )}
        </div>
        {editPass && (
          <div className="form-field">
            <label htmlFor="confirmPass"> Confirm Password: </label>
            <input
              type="password"
              name="confirmPass"
              value={confirmPass}
              onChange={(e) => {
                e.target.className = 'touched';
                setConfirmPass(e.target.value);
              }}
              onKeyDown={(e) => handleKeyDown(e)}
            />
            {confirmPassErr && <p className="error">{confirmPassErr}</p>}
          </div>
        )}
        <div className="form-field-check">
          <input
            type="checkbox"
            name="emailAlert"
            checked={emailAlert}
            onChange={() => setEmailAlert(!emailAlert)}
          />
          <label htmlFor="emailAlert">I want to receive email alerts.</label>
        </div>
        <div className="form-field-check">
          <input
            type="checkbox"
            name="textAlert"
            checked={textAlert}
            onChange={() => setTextAlert(!textAlert)}
          />
          <label htmlFor="textAlert">I want to receive text alerts.</label>
        </div>
        <div className="form-field">
          <label htmlFor="currentPass">Current Password: </label>
          <input
            type="password"
            name="currentPass"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
          {!posting && <button className="submit">Save</button>}
          {posting && (
            <div className="posting">
              <div className="spinner"></div>
            </div>
          )}
        </div>
        {postError && <p className="post-error">{postError}</p>}
        {posted && <p className="posted">Account updated successfully.</p>}
      </form>
    </section>
  );
};

export default AccountInfoForm;
