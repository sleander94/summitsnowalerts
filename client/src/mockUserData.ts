export const mockUserData = {
  _id: 'aoifh73$%D3g3',
  name: 'Test User',
  email: 'test@email.com',
  phone: '(555)555-5555',
  password: 'aorinferu142431e5e23e',
  notifications: {
    text: true,
    email: false,
    times: {
      0: true,
      1: false,
      2: false,
      3: false,
      4: false,
      5: false,
      6: false,
      7: false,
      8: true,
      9: false,
      10: false,
      11: false,
      12: false,
      13: false,
      14: false,
      15: false,
      16: false,
      17: false,
      18: true,
      19: false,
      20: false,
      21: false,
      22: false,
      23: false,
    },
    days: {
      sunday: true,
      monday: false,
      tuesday: false,
      wednesday: false,
      thursday: true,
      friday: true,
      saturday: true,
    },
  },
  mountains: { Breckenridge: 80424, Keystone: 80435 },
};
