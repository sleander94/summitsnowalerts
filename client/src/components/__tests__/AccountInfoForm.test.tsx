import AccountInfoForm from '../account-actions/AccountInfoForm';
import {
  render,
  cleanup,
  screen,
  waitFor,
  within,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { mockUserData } from '../../mockUserData';

const server = setupServer(
  rest.get('/users', (req, res, ctx) => {
    return res(ctx.json(mockUserData));
  }),
  rest.put('/users', (req, res, ctx) => {
    return res(ctx.status(400), ctx.json({ message: 'Bad Request.' }));
  })
);

const user = userEvent.setup();

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  cleanup;
});
afterAll(() => server.close());

it('Renders with correct user data displayed after fetch', async () => {
  const component = render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  await waitFor(() =>
    expect(screen.getByText('Test User')).toBeInTheDocument()
  );
  expect(screen.getByText('(555)555-5555')).toBeInTheDocument();
  expect(screen.getByText('test@email.com')).toBeInTheDocument();

  expect(component).toMatchSnapshot();
});

it('Displays name input field & saves user input', async () => {
  const component = render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  const field = await waitFor(() => screen.getByTestId('name'));
  expect(field).toBeInTheDocument();

  const { getByRole, queryByRole, getByText } = within(field);
  let button = getByRole('button');
  expect(button).toBeInTheDocument();

  await user.click(button);

  const input = getByRole('textbox');
  expect(input).toHaveValue('Test User');

  await user.clear(input);
  await user.type(input, 'New Name');
  expect(input).toHaveValue('New Name');

  button = getByRole('button');
  await user.click(button);
  expect(queryByRole('textbox')).not.toBeInTheDocument();
  expect(getByText('New Name')).toBeInTheDocument();
});

it('Displays error message when name field is empty', async () => {
  const component = render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  const field = await waitFor(() => screen.getByTestId('name'));

  const { getByRole, queryByRole, getByText } = within(field);

  let button = getByRole('button');
  await user.click(button);

  const input = getByRole('textbox');

  await user.clear(input);

  button = getByRole('button');
  await user.click(button);

  expect(queryByRole('textbox')).toBeInTheDocument();
  expect(getByText('Enter a name.')).toBeInTheDocument();
});

it('Displays email input field & saves user input', async () => {
  const component = render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  const field = await waitFor(() => screen.getByTestId('email'));
  expect(field).toBeInTheDocument();

  const { getByRole, queryByRole, getByText } = within(field);
  let button = getByRole('button');
  expect(button).toBeInTheDocument();

  await user.click(button);

  const input = getByRole('textbox');
  expect(input).toHaveValue('test@email.com');

  await user.clear(input);
  await user.type(input, 'coolemail@gmail.com');
  expect(input).toHaveValue('coolemail@gmail.com');

  button = getByRole('button');
  await user.click(button);
  expect(queryByRole('textbox')).not.toBeInTheDocument();
  expect(getByText('coolemail@gmail.com')).toBeInTheDocument();
});

it('Displays error message when email field is invalid', async () => {
  const component = render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  const field = await waitFor(() => screen.getByTestId('email'));

  const { getByRole, queryByRole, getByText } = within(field);

  let button = getByRole('button');
  await user.click(button);

  const input = getByRole('textbox');

  await user.clear(input);
  expect(getByText('Enter a valid email.')).toBeInTheDocument();

  await user.type(input, 'bademail.com');
  expect(getByText('Enter a valid email.')).toBeInTheDocument();

  button = getByRole('button');
  await user.click(button);

  expect(queryByRole('textbox')).toBeInTheDocument();
  expect(getByText('Enter a valid email.')).toBeInTheDocument();
});

it('Displays phone input field & saves user input', async () => {
  render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  const field = await waitFor(() => screen.getByTestId('phone'));
  expect(field).toBeInTheDocument();

  const { getByRole, queryByRole, getByText } = within(field);
  let button = getByRole('button');
  expect(button).toBeInTheDocument();

  await user.click(button);

  const input = getByRole('textbox');
  expect(input).toHaveValue('(555)555-5555');

  await user.clear(input);
  await user.type(input, '7191234567');
  expect(input).toHaveValue('7191234567');

  button = getByRole('button');
  await user.click(button);
  expect(queryByRole('textbox')).not.toBeInTheDocument();
  expect(getByText('7191234567')).toBeInTheDocument();
});

it('Displays error message when phone field is invalid', async () => {
  const component = render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  const field = await waitFor(() => screen.getByTestId('phone'));

  const { getByRole, queryByRole, getByText } = within(field);

  let button = getByRole('button');
  await user.click(button);

  const input = getByRole('textbox');

  await user.clear(input);
  await user.type(input, '5555555a');

  button = getByRole('button');
  await user.click(button);

  expect(queryByRole('textbox')).toBeInTheDocument();
  expect(getByText('Phone number cannot contain letters.')).toBeInTheDocument();
});

it('Displays new password & confirm password input fields & saves user input', async () => {
  render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  let button = await waitFor(() => screen.getByTestId('pass-button'));
  expect(button).toBeInTheDocument();

  await user.click(button);

  let passInput = screen.getByTestId('pass-input');
  let confirmPassInput = screen.getByTestId('confirm-pass-input');

  await user.type(passInput, 'goodpassword1');
  expect(passInput).toHaveValue('goodpassword1');

  await user.type(confirmPassInput, 'goodpassword1');
  expect(confirmPassInput).toHaveValue('goodpassword1');

  button = screen.getByTestId('pass-button');
  await user.click(button);
  expect(screen.getByText('*************')).toBeInTheDocument();
});

it('Displays error message when invalid new password is entered', async () => {
  render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  let button = await waitFor(() => screen.getByTestId('pass-button'));
  expect(button).toBeInTheDocument();

  await user.click(button);

  let passInput = screen.getByTestId('pass-input');
  let confirmPassInput = screen.getByTestId('confirm-pass-input');

  await user.type(passInput, 'short');
  await user.type(confirmPassInput, 'short');

  button = screen.getByTestId('pass-button');
  await user.click(button);

  expect(passInput).toBeInTheDocument();
  expect(confirmPassInput).toBeInTheDocument();
  expect(
    screen.getByText('6+ characters with 1 letter and 1 number.')
  ).toBeInTheDocument();
});

it('Displays error message when passwords do not match', async () => {
  render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  let button = await waitFor(() => screen.getByTestId('pass-button'));
  expect(button).toBeInTheDocument();

  await user.click(button);

  let passInput = screen.getByTestId('pass-input');
  let confirmPassInput = screen.getByTestId('confirm-pass-input');

  await user.type(passInput, 'password123');
  await user.type(confirmPassInput, 'password122');

  button = screen.getByTestId('pass-button');
  await user.click(button);

  expect(passInput).toBeInTheDocument();
  expect(confirmPassInput).toBeInTheDocument();
  expect(screen.getByText("Passwords don't match.")).toBeInTheDocument();
});

it('Displays error message on submission with invalid password', async () => {
  render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  let button = await waitFor(() => screen.getByText('Save'));

  await user.click(button);

  await waitFor(() =>
    expect(
      screen.getByText('Error updating account. Please try again.')
    ).toBeInTheDocument()
  );
});

it('Displays error message on submission with valid, but incorrect password', async () => {
  server.use(
    rest.put('/users', (req, res, ctx) => {
      return res(ctx.status(401), ctx.json({ message: 'Incorrect Password.' }));
    })
  );
  render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  let button = await waitFor(() => screen.getByText('Save'));

  await user.click(button);

  await waitFor(() =>
    expect(screen.getByText('Incorrect password.')).toBeInTheDocument()
  );
});

it('Displays confirmation message after successfully updating account', async () => {
  server.use(
    rest.put('/users', (req, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json({ message: 'Account updated successfully.' })
      );
    })
  );
  render(
    <Router>
      <AccountInfoForm user={mockUserData} />
    </Router>
  );

  let button = await waitFor(() => screen.getByText('Save'));

  await user.click(button);

  await waitFor(() =>
    expect(
      screen.getByText('Account updated successfully.')
    ).toBeInTheDocument()
  );
});
