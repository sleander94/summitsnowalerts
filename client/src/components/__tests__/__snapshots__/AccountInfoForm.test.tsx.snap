// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with correct user data displayed after fetch 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        id="account-info-form"
      >
        <div
          class="background-image"
        />
        <div
          class="content"
        >
          <h1>
            Account Info
          </h1>
          <form
            action=""
          >
            <div
              class="form-field"
              data-testid="name"
            >
              <label
                for="name"
              >
                Name: 
              </label>
              <p>
                Test User
              </p>
              <button
                type="button"
              >
                <img
                  alt="Edit"
                  src="icons8-edit.svg"
                />
              </button>
            </div>
            <div
              class="form-field"
              data-testid="email"
            >
              <label
                for="email"
              >
                Email: 
              </label>
              <p>
                test@email.com
              </p>
              <button
                type="button"
              >
                <img
                  alt="Edit"
                  src="icons8-edit.svg"
                />
              </button>
            </div>
            <div
              class="form-field"
              data-testid="phone"
            >
              <label
                for="phone"
              >
                Phone: 
              </label>
              <p>
                (555)555-5555
              </p>
              <button
                type="button"
              >
                <img
                  alt="Edit"
                  src="icons8-edit.svg"
                />
              </button>
            </div>
            <div
              class="form-field"
              data-testid="newPass"
            >
              <label
                for="newPass"
              >
                Password: 
              </label>
              <p>
                ******
              </p>
              <button
                data-testid="pass-button"
                type="button"
              >
                <img
                  alt="Edit"
                  src="icons8-edit.svg"
                />
              </button>
            </div>
            <div
              class="form-field"
              data-testid="currentPass"
            >
              <label
                for="currentPass"
              >
                Current Password: 
              </label>
              <input
                name="currentPass"
                type="password"
                value=""
              />
              <button
                class="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      id="account-info-form"
    >
      <div
        class="background-image"
      />
      <div
        class="content"
      >
        <h1>
          Account Info
        </h1>
        <form
          action=""
        >
          <div
            class="form-field"
            data-testid="name"
          >
            <label
              for="name"
            >
              Name: 
            </label>
            <p>
              Test User
            </p>
            <button
              type="button"
            >
              <img
                alt="Edit"
                src="icons8-edit.svg"
              />
            </button>
          </div>
          <div
            class="form-field"
            data-testid="email"
          >
            <label
              for="email"
            >
              Email: 
            </label>
            <p>
              test@email.com
            </p>
            <button
              type="button"
            >
              <img
                alt="Edit"
                src="icons8-edit.svg"
              />
            </button>
          </div>
          <div
            class="form-field"
            data-testid="phone"
          >
            <label
              for="phone"
            >
              Phone: 
            </label>
            <p>
              (555)555-5555
            </p>
            <button
              type="button"
            >
              <img
                alt="Edit"
                src="icons8-edit.svg"
              />
            </button>
          </div>
          <div
            class="form-field"
            data-testid="newPass"
          >
            <label
              for="newPass"
            >
              Password: 
            </label>
            <p>
              ******
            </p>
            <button
              data-testid="pass-button"
              type="button"
            >
              <img
                alt="Edit"
                src="icons8-edit.svg"
              />
            </button>
          </div>
          <div
            class="form-field"
            data-testid="currentPass"
          >
            <label
              for="currentPass"
            >
              Current Password: 
            </label>
            <input
              name="currentPass"
              type="password"
              value=""
            />
            <button
              class="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
